cmd/encryption-v1.go

```diff
@@ -1157,3 +1160,140 @@ func DecryptObjectInfo(info *ObjectInfo, headers http.Header) (encrypted bool, e
 	}
 	return
 }
+
+// The customer key in the header is used by the gateway for encryption in the case of
+// s3 gateway double encryption. A new client key is derived from the customer provided
+// key to be sent to the s3 backend for encryption at the backend.
+func deriveClientKey(header http.Header, headerName, bucket, object string) ([32]byte, error) {
+	var key [32]byte
+	clientKey, err := base64.StdEncoding.DecodeString(header.Get(headerName))
+	if err != nil || len(clientKey) != 32 { // The client key must be 256 bits long
+		return key, crypto.ErrInvalidCustomerKey
+	}
+	mac := hmac.New(sha256.New, clientKey)
+	mac.Write([]byte(crypto.SSEC.String()))
+	mac.Write([]byte(path.Join(bucket, object)))
+	mac.Sum(key[:0])
+	return key, nil
+}
+
+// extract encryption options for pass through to backend in the case of gateway
+func extractEncryptionOption(header http.Header, copySource bool, metadata map[string]string) (opts ObjectOptions, err error) {
+	var clientKey []byte
+	var sse encrypt.ServerSide
+
+	if copySource {
+		if crypto.SSECopy.IsRequested(header) {
+			clientKey, err = base64.StdEncoding.DecodeString(header.Get(crypto.SSECopyKey))
+			if err != nil || len(clientKey) != 32 { // The client key must be 256 bits long
+				return opts, crypto.ErrInvalidCustomerKey
+			}
+			if sse, err = encrypt.NewSSEC(clientKey); err != nil {
+				return
+			}
+			return ObjectOptions{ServerSideEncryption: encrypt.SSECopy(sse)}, nil
+		}
+		return
+	}
+
+	if crypto.SSEC.IsRequested(header) {
+		clientKey, err = base64.StdEncoding.DecodeString(header.Get(crypto.SSECKey))
+		if err != nil || len(clientKey) != 32 { // The client key must be 256 bits long
+			return opts, crypto.ErrInvalidCustomerKey
+		}
+		if sse, err = encrypt.NewSSEC(clientKey); err != nil {
+			return
+		}
+		return ObjectOptions{ServerSideEncryption: sse}, nil
+	}
+	if crypto.S3.IsRequested(header) || (metadata != nil && crypto.S3.IsEncrypted(metadata)) {
+		return ObjectOptions{ServerSideEncryption: encrypt.NewSSE()}, nil
+	}
+	return opts, nil
+}
+
+// get ObjectOptions for GET calls from encryption headers
+func getEncryptionOpts(r *http.Request, bucket, object string) (ObjectOptions, error) {
+	var (
+		encryption encrypt.ServerSide
+		opts       ObjectOptions
+	)
+	if GlobalGatewaySSE != nil {
+		for _, k := range GlobalGatewaySSE {
+			if k == GatewaySSEC && crypto.SSEC.IsRequested(r.Header) {
+				derivedKey, err := deriveClientKey(r.Header, crypto.SSECKey, bucket, object)
+				if err != nil {
+					return opts, err
+				}
+				encryption, err = encrypt.NewSSEC(derivedKey[:])
+				if err != nil {
+					return opts, err
```

Comment:
Probably `logger.Criticalif` - this situations can never happen b/c either the client key was malformed (should have failed before) or the client key is well-formed, so should be the derived key.

Reply:
good point - changed this.